### PR TITLE
homogenize fake discovery

### LINKED ISSUE #62031: Too many defs of FakeDiscovery: Needs to be homogenized
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug
> /kind feature

/kind cleanup

**What happened**:

Hi Meows

Currently we have several independent defs of `fake discovery` which is confusing and hard to read.

As is shown below, a list of dup defs:

- [pkg/kubectl/cmd/top_pod_test.go](https://github.com/kubernetes/kubernetes/blob/df3637967015c9770c67140d4bbf99467c87abf8/pkg/kubectl/cmd/top_pod_test.go#L382)
- [staging/src/k8s.io/client-go/discovery/cached/memcache_test.go](https://github.com/kubernetes/kubernetes/blob/3f0141edd52ab6fd7b6fe473ecceb480faf12079/staging/src/k8s.io/client-go/discovery/cached/memcache_test.go#L29)
- [staging/src/k8s.io/client-go/discovery/fake/discovery.go](https://github.com/kubernetes/kubernetes/blob/a9e244d81f05d1fa6965f33b9a3ec04cce6d058f/staging/src/k8s.io/client-go/discovery/fake/discovery.go) *SHOULD* be the one and only

**What you expected to happen**:

Only one defs of `fake discovery`

### PR DESCRIPTION
**Pull request by yue9944882**
**What this PR does / why we need it**:

Cleans fake discovery and leaves the only one defined in `client-go/discovery`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #62031

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by yue9944882** [comment]
/cc @liggitt Could u shed some light on this? :) 
Since you are in the both OWNERS file.

**Comment by wenjiaswe** [comment]
cc @yliaog

File: staging/src/k8s.io/client-go/discovery/fake/discovery.go [hunk 1]
```diff
@@ -51,11 +51,6 @@ func (c *FakeDiscovery) ServerResourcesForGroupVersion(groupVersion string) (*me
 }
 
 func (c *FakeDiscovery) ServerResources() ([]*metav1.APIResourceList, error) {
-	action := testing.ActionImpl{
-		Verb:     "get",
-		Resource: schema.GroupVersionResource{Resource: "resource"},
-	}
-	c.Invokes(action, nil)
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/discovery/fake/discovery.go]: what's the reason for the change here?

Comment (yue9944882) [reply staging/src/k8s.io/client-go/discovery/fake/discovery.go]: @yliaog I believe these codes are useless here, aren't they? Am I missing sth?

Comment (yliaog) [reply staging/src/k8s.io/client-go/discovery/fake/discovery.go]: i tend to agree with you, however, i don't quite understand why the code for c.Invoke(...) was introduced in the first place, similarly for the other methods ServerResourcesForGroupVersion(), ServerGroups(), ServerVersion()

Comment (yue9944882) [reply staging/src/k8s.io/client-go/discovery/fake/discovery.go]: Yeah.. IDK either. But I'm pretty sure it should be cleaned up anyway.

Comment (yue9944882) [reply staging/src/k8s.io/client-go/discovery/fake/discovery.go]: Cleaning up for for other methods too :)

File: staging/src/k8s.io/client-go/discovery/fake/discovery.go [hunk 2]
```diff
@@ -37,11 +37,6 @@ type FakeDiscovery struct {
 }
```
Comment (yliaog) [inline staging/src/k8s.io/client-go/discovery/fake/discovery.go]: could you update the comment above the type, that mentioned testing.Fake.Invoke

**Comment by yue9944882** [comment]
@yliaog I just rebased to fix the bazel build failure. Could you please take another look at this once it's all green in the CI? :)

**Comment by yliaog** [comment]
Sure

**Comment by yue9944882** [comment]
@yliaog PTAL thx :)

**Comment by soltysh** [comment]
Please squash your changes into a single commit then I'll approve the kubectl bits, you still need approval for the client-go, though.

**Comment by yue9944882** [comment]
@soltysh Thanks, PTAL

**Comment by deads2k** [comment]
We need to be able to track actions.  There are two reasons for using a mock:
 1. to get back a return value (you've solved this I think)
 2. to make sure that your code is calling the correct client methods.

this pull eliminates the ability to do item 2.

File: staging/src/k8s.io/client-go/discovery/fake/discovery.go [hunk 3]
```diff
@@ -68,12 +58,6 @@ func (c *FakeDiscovery) ServerPreferredNamespacedResources() ([]*metav1.APIResou
...
 
 func (c *FakeDiscovery) ServerGroups() (*metav1.APIGroupList, error) {
-	action := testing.ActionImpl{
-		Verb:     "get",
-		Resource: schema.GroupVersionResource{Resource: "group"},
-	}
-	c.Invokes(action, nil)
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/fake/discovery.go]: I think we need this or something very much like it.

**Comment by yue9944882** [comment]
> 2. to make sure that your code is calling the correct client methods.

@deads2k Thanks. But it somehow still looks confusing to me THO, because these code might only make a difference by panics  when `*testing.Fake` is nil.

I'll revert the changes of this part.

**Comment by yue9944882** [comment]
@deads2k Friendly ping for a review 😃
